operator/v1/types_console.go

```diff
@@ -202,6 +205,30 @@ type AddPage struct {
 	DisabledActions []string `json:"disabledActions,omitempty"`
 }
 
+type PerspectiveState string
+
+// These are valid perspective states. "Enabled" means the perspective is shown.
+// "Disabled" means the Perspective is hidden.
+const (
+	Enabled  PerspectiveState = "True"
+	Disabled PerspectiveState = "False"
+)
+
+type Perspectives struct {
+	// id defines the id of perspective to disable.
+	// Incorrect or unknown ids will be ignored.
+	// +kubebuilder:validation:Required
+	ID string `json:"id"`
+	// requiresAccessReview defines a list of permission checks.
+	// The perspective will only be shown when all checks are successful.
+	// +optional
+	RequiresAccessReview []authorizationv1.ResourceAttributes `json:"requiresAccessReview,omitempty"`
```

Comment:
We tend to discourage using external types because they aren't under our control. They may have additional fields that you don't need within your API or, over time, they may add additional fields which suddenly get added to your API.

My advice would be to copy the ResourceAttributes here if you want to use a similar structure. This also allows you to customise the comments (and therefore generated docs) to your specific API use case

Reply:
That is a good point, but we should also keep in mind this is a versioned k8s API (k8s.io/api/authorization/v1), and we use `ResourceAttributes` also in `types_console_quick_start.go` for a similar case:

https://github.com/openshift/api/blob/0b39f81154fa42abe26933181141ed4b4e3ad247/console/v1/types_console_quick_start.go#L70-L74

We would also pass the object 1:1 to an RBAC check. So for me, it makes sense to use than the k8sn API which also supports things like subresources etc. which we might not need for our case. But we want to develop this feature in a (generic) way so that customers can apply their own rules (permission checks) if needed. If we create our own type that supports only resource and verb, this will add IMO unnecessary limits.

Reply:
When a value is interpreted by the consumer of the API, then precise control of the schema is a requirement since the specification of additional fields into an existing v1 API (fully allowed), can cause mishandling of fields and defy user expectation.  For instance, imagine we add another field about storage version to the access review.  If you had code reading and interpreting the value, then the semantics of the field you're referencing would be out of date and a user that expects a reject may suddenly get an allow.

When a value is used fully opaquely and passed through to a matching endpoint, a case can be made to use a type that has a stability guarantee equal to or greater than the stability level used in OCP.

Reply:
Agree. We pass it through an endpoint that understands auth-checks. We don't interpret these permission checks on our own. So I would expect we're fine here.

Reply:
I checked the detail now, we send it to k8s `group: authorization.k8s.io/v1`, `kind: SelfSubjectAccessReview` that get this a (single) `ResourceAttributes`.